stetho/src/main/java/com/facebook/stetho/inspector/elements/Document.java

```diff
@@ -0,0 +1,473 @@
+/*
+ * Copyright (c) 2014-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+package com.facebook.stetho.inspector.elements;
+
+import android.os.SystemClock;
+import android.support.annotation.Nullable;
+import com.facebook.stetho.common.Accumulator;
+import com.facebook.stetho.common.ArrayListAccumulator;
+import com.facebook.stetho.common.LogUtil;
+import com.facebook.stetho.common.UncheckedCallable;
+import com.facebook.stetho.inspector.helper.ObjectIdMapper;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Queue;
+
+public class Document {
+  private DOMProvider.Factory mFactory;
+
+  @Nullable
+  private DOMProvider mDOMProvider;
+  private ShadowDOM mShadowDOM;
+  private ObjectIdMapper mObjectIdMapper;
+  private UpdateListener mUpdateListener;
+
+  private AttributeListAccumulator mCachedAttributeAccumulator;
+  private ChildEventingList mCachedChildEventingList;
+  private ArrayListAccumulator<Object> mCachedChildrenAccumulator;
+
+  public Document(DOMProvider.Factory factory) {
+    this.mFactory = factory;
+    this.mObjectIdMapper = new DOMObjectIdMapper();
+  }
+
+  // dom provider should be created only once, and if multiple
+  // domains use the class Document, errors could arise if this
+  // was merely a `createDomProvider` method
+  public void createDomProviderIfNoneExists() {
```

Comment:
Uhm, what is the `IfNoneExists` part of this?  I don't see any such enforcement.